Interpret __builtin_unreachable() as no-op for sv-comp

fbrausse opened the pull request:
Fixes a part of #1494:

	Parsing ../../sv-benchmarks/c/coreutils-v8.31/env_3args_ok.c
	../../sv-benchmarks/c/coreutils-v8.31/env_3args_ok.c:48729:6: error: expected identifier or '('
	void __builtin_unreachable();
	     ^
	<command line>:5:33: note: expanded from here
	#define __builtin_unreachable() 1

I'll keep this a draft until an sv-comp run has confirmed I'm not breaking anything.

fbrausse committed:
[svcomp] wrapper: conditionally add #define of __builtin_unreachable()

Fixes a part of #1494:

	Parsing ../../sv-benchmarks/c/coreutils-v8.31/env_3args_ok.c
	../../sv-benchmarks/c/coreutils-v8.31/env_3args_ok.c:48729:6: error: expected identifier or '('
	void __builtin_unreachable();
	     ^
	<command line>:5:33: note: expanded from here
	#define __builtin_unreachable() 1

fbrausse committed:
[svcomp] wrapper: regexify pthread_create

mikhailramalho commented:
I wonder if we could just handle the __builtin_unrechable() as abort(). I could easily support it on my open PR, if I get to finish it in time.

fbrausse commented:
> I wonder if we could just handle the __builtin_unrechable() as abort(). I could easily support it on my open PR, if I get to finish it in time.

In https://gitlab.com/sosy-lab/benchmarking/sv-benchmarks/-/issues/1296 they argue about the soundness of it and it seems the conclusion is to not give it any semantics, that's why I thought it best to get rid of it whereever we can.

fbrausse commented:
> I wonder if we could just handle the __builtin_unrechable() as abort(). I could easily support it on my open PR, [...]

For the non-sv-comp version of ESBMC, I totally support handling __builtin_unreachable() like abort().

fbrausse commented:
> handling __builtin_unreachable() like abort().

I think we already do: https://github.com/esbmc/esbmc/blob/face86b64ac1c3bea28ecbfd7ff2d895ef7355a4/src/goto-programs/builtin_functions.cpp#L572-L573

mikhailramalho commented:
right, so something like:

#ifdef svcomp
if(base_name == "__builtin_unreachable")
  return;
#endif

in there should be enough for the competition

Em qui., 16 de nov. de 2023 às 13:16, Franz Brauße ***@***.***>
escreveu:

> handling __builtin_unreachable() like abort().
>
> I think we already do:
> https://github.com/esbmc/esbmc/blob/face86b64ac1c3bea28ecbfd7ff2d895ef7355a4/src/goto-programs/builtin_functions.cpp#L572-L573
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/esbmc/esbmc/pull/1500#issuecomment-1814776428>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKEJH6BHHJIWEHWNOFZ26TYEY343AVCNFSM6AAAAAA7OBJ6HOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTQMJUG43TMNBSHA>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 

Mikhail R. Gadelha, Ph.D.

fbrausse commented:
Good point, I'll try that. Looks much cleaner.

fbrausse committed:
Revert "[svcomp] wrapper: regexify pthread_create"

This reverts commit dc4fbb0887992cf980a3388e4dd5136240e26884.

fbrausse committed:
Revert "[svcomp] wrapper: conditionally add #define of __builtin_unreachable()"

This reverts commit d0a3231b8fe2c7a0def799761986ffc149624879.

fbrausse committed:
[goto] builtin: for SV-COMP interpret __builtin_unreachable() as a no-op

fbrausse committed:
[svcomp] wrapper: do not #define __builtin_unreachable()

This leads to problems:

	Parsing ../../sv-benchmarks/c/coreutils-v8.31/env_3args_ok.c
	../../sv-benchmarks/c/coreutils-v8.31/env_3args_ok.c:48729:6: error: expected identifier or '('
	void __builtin_unreachable();
	     ^
	<command line>:5:33: note: expanded from here
	#define __builtin_unreachable() 1

Note, this affects the verdict of

	c/ldv-linux-4.0-rc1-mav/linux-4.0-rc1---drivers--mtd--devices--docg3.ko.cil.i

mikhailramalho reviewed (approved):
Lgtm

lucasccordeiro reviewed (approved):
Thanks for submitting this PR, @fbrausse.
